packages/astro/src/assets/services/service.ts

```diff
@@ -0,0 +1,155 @@
+	 */
+	parseURL: (url: URL) => Record<string, any>;
+	/**
+	 * Performs the image transformations on the input image and returns both the binary data and
+	 * final image format of the optimized image.
+	 *
+	 * @param inputBuffer Binary buffer containing the original image.
+	 * @param transform @type {TransformOptions} defining the requested transformations.
+	 */
+	transform: (
+		inputBuffer: Buffer,
+		transform: Record<string, any>
+	) => Promise<{ data: Buffer; format: OutputFormat }>;
+}
+
+/**
+ * Basic local service to take things from
+ */
+export const baseService: Omit<LocalImageService, 'transform'> = {
+	getHTMLAttributes(options) {
+		let targetWidth = options.width;
+		let targetHeight = options.height;
+		if (isESMImportedImage(options.src)) {
+			const aspectRatio = options.src.width / options.src.height;
+
+			// If we have a desired height and no width, calculate the target width automatically
+			if (targetHeight && !targetWidth) {
+				targetWidth = Math.round(targetHeight * aspectRatio);
+			} else if (targetWidth && !targetHeight) {
+				targetHeight = Math.round(targetWidth / aspectRatio);
+			}
+		}
+
+		const { src, width, height, format, quality, ...attributes } = options;
+
+		return {
+			...attributes,
+			width: targetWidth,
+			height: targetHeight,
+			loading: attributes.loading ?? 'lazy',
+			decoding: attributes.decoding ?? 'async',
+		};
+	},
+	getURL(options: ImageTransform) {
+		if (!isESMImportedImage(options.src)) {
+			// For non-ESM imported images, width and height are required to avoid CLS, as we can't infer them from the file
+			let missingDimension: 'width' | 'height' | 'both' | undefined;
+			if (!options.width && !options.height) {
+				missingDimension = 'both';
+			} else if (!options.width && options.height) {
+				missingDimension = 'width';
+			} else if (options.width && !options.height) {
+				missingDimension = 'height';
+			}
+
+			if (missingDimension) {
+				throw new AstroError({
+					...AstroErrorData.MissingImageDimension,
+					message: AstroErrorData.MissingImageDimension.message(missingDimension),
+				});
+			}
+		}
+
+		// Both our currently available local services don't handle remote images, so for them we can just return as is
+		if (!isESMImportedImage(options.src) && isRemotePath(options.src)) {
+			return options.src;
+		}
+
+		const searchParams = new URLSearchParams();
+		searchParams.append('href', isESMImportedImage(options.src) ? options.src.src : options.src);
+
+		options.width && searchParams.append('w', options.width.toString());
+		options.height && searchParams.append('h', options.height.toString());
+		options.quality && searchParams.append('q', options.quality.toString());
+		options.format && searchParams.append('f', options.format);
+
+		return '/_image?' + searchParams;
+	},
+	parseURL(url) {
+		const params = url.searchParams;
+
+		if (!params.has('href')) {
+			return {};
+		}
+
+		let transform: Record<string, any> = { src: params.get('href')! };
```

Comment:
Should this object have more of a defined shape? If so you could initialize the other values as `undefined`. This will also help with perf (if this function is called often enough for that to matter). Disregard if it shouldn't have more of a shape.